File: sdks/python/dcp_ai/v2/composite_sig.py
"""
Composite signature type definitions for DCP v2 hybrid classical/PQ signing.
"""

from __future__ import annotations

from typing import Literal, TypedDict


class SignatureEntry(TypedDict):
    alg: str
    kid: str
    sig_b64: str


class CompositeSignature(TypedDict):
    classical: SignatureEntry
    pq: SignatureEntry | None
    binding: Literal["pq_over_classical", "classical_only"]

File: sdks/typescript/src/core/composite-sig.ts
/**
 * DCP v2.0 Composite Signature types.
 *
 * Composite signatures cryptographically bind classical and PQ signatures
 * to prevent stripping attacks. The PQ signature covers the classical
 * signature, so removing either component breaks verification.
 */

export type BindingMode = 'pq_over_classical' | 'classical_only';

export interface SignatureEntry {
  alg: string;
  kid: string;
  sig_b64: string;
}

export interface CompositeSignature {
  classical: SignatureEntry;
  pq: SignatureEntry | null;
  binding: BindingMode;
}
